include/envoy/filesystem/filesystem.h

```diff
@@ -14,58 +14,62 @@ namespace Envoy {
 namespace Filesystem {
 
 /**
- * Abstraction for a file on disk.
+ * Abstraction for a basic file on disk.
  */
-class File {
+class RawFile {
 public:
-  virtual ~File() {}
+  virtual ~RawFile() {}
 
   /**
-   * Write data to the file.
+   * Open the file with O_RDWR | O_APPEND | O_CREAT
+   * The file will be closed when this object is destructed
+   *
+   * @return bool whether the open succeeded
    */
-  virtual void write(absl::string_view) PURE;
+  virtual Api::SysCallBoolResult open() PURE;
 
   /**
-   * Reopen the file.
+   * Write the buffer to the file. The file must be explicitly opened before writing.
+   *
+   * @return ssize_t number of bytes written, or -1 for failure
    */
-  virtual void reopen() PURE;
+  virtual Api::SysCallSizeResult write(absl::string_view buffer) PURE;
 
   /**
-   * Synchronously flush all pending data to disk.
+   * Close the file.
+   *
+   * @return bool whether the close succeeded
    */
-  virtual void flush() PURE;
-};
+  virtual Api::SysCallBoolResult close() PURE;
 
-typedef std::shared_ptr<File> FileSharedPtr;
+  /**
+   * @return bool is the file open
+   */
+  virtual bool isOpen() PURE;
 
-/**
- * Captures state, properties, and stats of a file-system.
- */
-class Instance {
-public:
-  virtual ~Instance() {}
+  /**
+   * @return string the file path
+   */
+  virtual std::string path() PURE;
 
   /**
-   * Creates a file, overriding the flush-interval set in the class.
-   *
-   * @param path The path of the file to open.
-   * @param dispatcher The dispatcher used for set up timers to run flush().
-   * @param lock The lock.
-   * @param file_flush_interval_msec Number of milliseconds to delay before flushing.
+   * @return string a human-readable string describing the error code
+   * TODO(sesmith177) Abstract this method so it isn't dependant on integer error codes
    */
-  virtual FileSharedPtr createFile(const std::string& path, Event::Dispatcher& dispatcher,
-                                   Thread::BasicLockable& lock,
-                                   std::chrono::milliseconds file_flush_interval_msec) PURE;
+  virtual std::string errorToString(int error) PURE;
+};
+
+typedef std::unique_ptr<RawFile> RawFilePtr;
+
+class RawInstance {
```

Comment:
It seems a little strange to me that we have a "raw" filesystem instance and then a "normal" filesystem instance. Why not just have a single filesystem that can create either "raw" or "regular" files? (Final names TBD).

Reply:
The issue is that creating the "regular" file requires a Dispatcher, a lock, stats, and threading. In the final version of this (with Windows support), we want to be able to inject the platform specific filesystem code from here: https://github.com/envoyproxy/envoy/blob/f9107b26ccca409a13716b7b094bd87fec9765fb/source/exe/win32/platform_impl.h, similarly to how we inject the platform specific threading code.

To get around this, the "raw" filesystem instance doesn't know how to create "regular" files -- it is injected to the "normal" instance that does.

If this too strange, then maybe just the Api object will be able to create "regular" files (since they are not platform specific). This would end up reverting a bit of #5692, which wouldn't be a problem, except for the small bit of churn

Reply:
See my comment above. Should we just have "regular" files be part of the access log manager interface since this is the interface the reopens them, etc.?